politeiawww/proposals.go

```diff
@@ -2275,39 +2741,327 @@ func (p *politeiawww) processStartVoteV2(sv www2.StartVote, u *user.User) (*www2
+func voteIsApproved(vs www.VoteSummary) bool {
+	if vs.Status != www.PropVoteStatusFinished {
+		// Vote has not ended yet
+		return false
+	}
 
-	bb, err := p.getBestBlock()
+	var (
+		total   uint64
+		approve uint64
+	)
+	for _, v := range vs.Results {
+		total += v.VotesReceived
+		if v.Option.Id == decredplugin.VoteOptionIDApprove {
+			approve = v.VotesReceived
+		}
+	}
+	quorum := uint64(float64(vs.QuorumPercentage) / 100 * float64(vs.EligibleTickets))
+	pass := uint64(float64(vs.PassPercentage) / 100 * float64(total))
+	switch {
+	case total < quorum:
+		// Quorum not met
+		return false
+	case approve < pass:
+		// Pass percentage not met
+		return false
+	}
+
+	return true
+}
+
+// processStartVoteRunoffV2 starts the runoff voting process on all public,
+// non-abandoned RFP submissions for the provided RFP token. If politeiad fails
+// to start the voting period on any of the RFP submissions, all work is
+// unwound and an error is returned.
+func (p *politeiawww) processStartVoteRunoffV2(sv www2.StartVoteRunoff, u *user.User) (*www2.StartVoteRunoffReply, error) {
+	log.Tracef("processStartVoteRFP %v", sv.Token)
+
+	// Sanity check
+	if !u.Admin {
+		return nil, fmt.Errorf("user is not an admin")
+	}
+
+	// Validate the StartVotes. The Token and ProposalVersion are
+	// validated later in this function once we've fetched all of
+	// the rfp submission proposals.
+	for _, v := range sv.StartVotes {
+		err := validateStartVoteV2(v, p.cfg.VoteDurationMin,
+			p.cfg.VoteDurationMax)
+		if err != nil {
+			return nil, err
+		}
+
+		// Ensure admin signed using their active identity
+		if v.PublicKey != u.PublicKey() {
+			return nil, www.UserError{
+				ErrorCode: www.ErrorStatusInvalidSigningKey,
+			}
+		}
+
+		// Vote options must be approve/reject and nothing else
+		err = validateVoteOptions(v.Vote.Options)
+		if err != nil {
+			return nil, err
+		}
+
+		// Vote type must be a runoff vote
+		if v.Vote.Type != www2.VoteTypeRunoff {
+			e := fmt.Sprintf("vote type must be %v", www2.VoteTypeRunoff)
+			return nil, www.UserError{
+				ErrorCode:    www.ErrorStatusInvalidVoteType,
+				ErrorContext: []string{e},
+			}
+		}
+	}
+
+	// Validate the RFP proposal
+	rfp, err := p.getProp(sv.Token)
+	if err != nil {
+		if err == cache.ErrRecordNotFound {
+			err = www.UserError{
+				ErrorCode:    www.ErrorStatusProposalNotFound,
+				ErrorContext: []string{sv.Token},
+			}
+		}
+		return nil, err
+	}
+	bestBlock, err := p.getBestBlock()
+	if err != nil {
+		return nil, err
+	}
+	rfpVoteSummary, err := p.voteSummaryGet(sv.Token, bestBlock)
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case !voteIsApproved(*rfpVoteSummary):
+		// The RFP submissions can only be voted on if the RFP
+		// proposal itself has been voted on and approved.
+		return nil, www.UserError{
+			ErrorCode:    www.ErrorStatusWrongVoteResult,
+			ErrorContext: []string{"rfp proposal vote did not pass"},
+		}
+	case rfp.LinkBy < time.Now().Unix():
```

Comment:
@lukebp 
Shouldn't this condition be the other way around ?
Currently if the deadline has expired this condition is `true` therefore the error is being thrown, I guess it should be the other way around, when deadline hasn't   `case time.Now().Unix() < rfp.LinkBy`  => throw error!

Reply:
Oh yeah, you're right.